lib/bundler/metrics.rb

```diff
@@ -0,0 +1,141 @@
+      end
+      options = Bundler.settings.all.join(",")
+      @metrics_hash["options"] = options unless options.empty?
+      @metrics_hash["ci"] = cis.join(",") if cis.any?
+      # add any user agent strings set in the config
+      extra_ua = Bundler.settings[:user_agent]
+      @metrics_hash["extra_ua"] = extra_ua if extra_ua
+      add_additional_metrics
+      write_to_file
+    end
+
+    # cant use it for now 
+    # def add_performance_metrics(time)
+    #   @metrics_hash["command_time_taken"] = time.round(2)
+    # end
+
+    def send_metrics
+      # dummy server for now
+      begin
+        uri = URI.parse("https://webhook.site/ee1e4493-c0f0-4e40-84fb-3a36d79d47fb")
+        http = Net::HTTP.new(uri.host, uri.port)
+        http.use_ssl = true
+        request = Net::HTTP::Post.new(uri.request_uri)
+        request.set_form_data(read_from_file)
+        http.request(request)
+      rescue SocketError
+        "TCP connection failed"
+      end
+
+      # We've sent the metrics so we empty the file
+      # File::TRUNC is preferable since File.truncate doesn't work for all systems
+      open(@path, File::TRUNC) if File.exist?(@path)
+    end
+
+  private
+
+    def write_to_file
+      SharedHelpers.filesystem_access(@path) do |file|
+        FileUtils.mkdir_p(file.dirname) unless File.exist?(file)
+        require "psych"
+        File.open(file, "a") {|f| f.write(Psych.dump(@metrics_hash)) }
+      end
+    end
+
+    def read_from_file
+      valid_file = @path.exist? && !@path.size.zero?
+      return {} unless valid_file
+      list = Array.new
+      SharedHelpers.filesystem_access(@path, :read) do |file|
+        require "psych"
+        Psych.load_stream(file.read) {|doc| list << doc }
+      end
+      list
+    end
+
+    def cis
+      env_cis = {
+        "TRAVIS" => "travis",
+        "CIRCLECI" => "circle",
+        "SEMAPHORE" => "semaphore",
+        "JENKINS_URL" => "jenkins",
+        "BUILDBOX" => "buildbox",
+        "GO_SERVER_URL" => "go",
+        "SNAP_CI" => "snap",
+        "CI_NAME" => ENV["CI_NAME"],
+        "CI" => "ci",
+      }
+      env_cis.find_all {|env, _| ENV[env] }.map {|_, ci| ci }
+    end
+
+    def add_additional_metrics
+      case ARGV.first
+      when "install"
+        @metrics_hash["gemfile_gem_count"] = Bundler.definition.dependencies.count
+        @metrics_hash["installed_gem_count"] = Bundler.definition.specs.count
+        @metrics_hash["git_gem_count"] = Bundler.definition.sources.git_sources.count
+        @metrics_hash["path_gem_count"] = Bundler.definition.sources.path_sources.count
+        @metrics_hash["rubygems_source_count"] = Bundler.definition.sources.rubygems_sources.count
+        @metrics_hash["gem_sources"] = Bundler.definition.sources.rubygems_sources.map {|s| Digest::MD5.hexdigest(s.get_source) }
+      when "exec"
+        ARGV[1..-1].each_index {|i| @metrics_hash["executed_command_#{i+1}"] = ARGV[i+1] }
```

Comment:
I think we will want to record some sort of information as part of exec, but I don't think we will want to record the commands that people are running. Sometimes those commands contain private information, and we don't want to send private information to the server. Let's remove `ARGV` from here, and instead try to work out how to time how long it takes Bundler to load and do everything before it runs `Kernel.exec` or `Kernel.load`.

Reply:
I assume we push that to the queue that's created upon initialization now.
Do we also want to add the command (exec) to the hash that's pushed into the queue to differentiate between other single-command metrics we will add later?
ex `[{"timestamp" = ..., "command" = "exec", "time_taken" = "wx.yz"}]`

Additionally, I guess we will want to append the queue to the metrics.yml file, and then send it alongside the system info?

Reply:
yes, I think we will want to include the command, and append to the queue, to be sent later.